Requests with ReactiveElasticsearchRepository methods doesn't fail if it can't connect with Elasticsearch

**Spring Boot:** 2.3.9.RELEASE
**Spring Data Elastic Search:** 4.0.7.RELEASE

## Problem

When doing a search request with a ReactiveElasticsearchRepository method, as for example `findAllById()`, if there's no connection with Elasticsearch, for example the cluster is stopped just before the request is made, the connection error is silenced and the request finish correctly but without returning values.


How to reproduce it:


1. With an Elasticsearch repository prepare a request like:

```java
        List<TestDocument> documents = repository.findAllById(Arrays.asList(ID1, ID2))
                .collectList()
                .block();
```

2. Set a breakpoint just before executing the search.
3. Start the service in debug mode with Elasticsearch running.
4. Once it stops in the breakpoint stop Elasticsearch.
5. With Elasticsearch stopped resume the execution.

*Result:* The request won't fail and an empty list will be returned.


I wasn't able to prepare a test as I couldn't pinpoint exactly where the problem occurs, but I prepared a project with this example [TestBug.test()](https://github.com/josemalonsom/spring-data-elasticsearch-bug/blob/a396c9e05108daa6f6980ef82136356e212d676c/src/test/java/com/example/TestBug.java#L35), it's a little clumsy, it requires to stop the execution with a breakpoint, sorry for that, but is the only way I found to reproduce it, I hope it helps.

I can reproduce this with you demo like you described. I still have to investigate why the dropped connection is not reported.
When changing your demo to use Spring Boot 2.4.3 (with Spring Data Elasticsearch 4.1.5), I get an error:

```
2021-03-01 08:07:01.398  WARN 11460 --- [ctor-http-nio-2] r.netty.http.client.HttpClientConnect    : [id: 0x2391c6d5, L:/127.0.0.1:57551 - R:localhost/127.0.0.1:9200] The connection observed an error

io.netty.handler.timeout.ReadTimeoutException: null

2021-03-01 08:07:01.409 ERROR 11460 --- [ctor-http-nio-2] reactor.core.publisher.Operators         : Operator called default onErrorDropped

reactor.core.Exceptions$ErrorCallbackNotImplemented: org.springframework.data.elasticsearch.client.NoReachableHostException: Host 'localhost:9200' not reachable. Cluster state is offline.
Caused by: org.springframework.data.elasticsearch.client.NoReachableHostException: Host 'localhost:9200' not reachable. Cluster state is offline.
	at org.springframework.data.elasticsearch.client.reactive.SingleNodeHostProvider.lambda$lookupActiveHost$4(SingleNodeHostProvider.java:108) ~[spring-data-elasticsearch-4.1.5.jar:4.1.5]
```

Is it an option to upgrade your application to the actual Spring Boot version?

I do not know currently if it is a change in Spring Data Elasticsearch or in Spring webflux that now shows the correct behaviour, I'll have to investigate that.

--- a/src/main/java/org/springframework/data/elasticsearch/client/reactive/DefaultReactiveElasticsearchClient.java
+++ b/src/main/java/org/springframework/data/elasticsearch/client/reactive/DefaultReactiveElasticsearchClient.java
@@ -866,6 +866,10 @@ public class DefaultReactiveElasticsearchClient implements ReactiveElasticsearch
 		String mediaType = response.headers().contentType().map(MediaType::toString).orElse(XContentType.JSON.mediaType());
 
 		return response.body(BodyExtractors.toMono(byte[].class)) //
+				.switchIfEmpty(Mono
+						.error(new ElasticsearchStatusException(String.format("%s request to %s returned error code %s and no body.",
+								request.getMethod(), request.getEndpoint(), statusCode), status))
+				)
 				.map(bytes -> new String(bytes, StandardCharsets.UTF_8)) //
 				.flatMap(content -> contentOrError(content, mediaType, status))
 				.flatMap(unused -> Mono

--- a/src/test/java/org/springframework/data/elasticsearch/client/reactive/DefaultReactiveElasticsearchClientTest.java
+++ b/src/test/java/org/springframework/data/elasticsearch/client/reactive/DefaultReactiveElasticsearchClientTest.java
@@ -22,20 +22,29 @@ import static org.mockito.Mockito.*;
 import reactor.core.publisher.Mono;
 import reactor.test.StepVerifier;
 
+import java.net.URI;
+import java.util.Optional;
 import java.util.function.Function;
 
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.action.get.GetRequest;
 import org.elasticsearch.action.search.SearchRequest;
 import org.elasticsearch.client.Request;
 import org.elasticsearch.index.query.QueryBuilders;
 import org.elasticsearch.search.builder.SearchSourceBuilder;
 import org.elasticsearch.search.fetch.subphase.FetchSourceContext;
-import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
+import org.mockito.Spy;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.reactive.function.client.ClientResponse;
+import org.springframework.web.reactive.function.client.WebClient;
 import org.springframework.web.reactive.function.client.WebClient.ResponseSpec;
+import org.springframework.web.util.UriBuilder;
 
 /**
  * @author Peter-Josef Meisch
@@ -46,29 +55,23 @@ class DefaultReactiveElasticsearchClientTest {
 	@Mock private HostProvider hostProvider;
 
 	@Mock private Function<SearchRequest, Request> searchRequestConverter;
+	@Spy private RequestCreator requestCreator;
 
-	private DefaultReactiveElasticsearchClient client;
+	@Mock private WebClient webClient;
 
-	@BeforeEach
-	void setUp() {
-		client = new DefaultReactiveElasticsearchClient(hostProvider, new RequestCreator() {
-			@Override
-			public Function<SearchRequest, Request> search() {
-				return searchRequestConverter;
-			}
-		}) {
+	@Test
+	void shouldSetAppropriateRequestParametersOnCount() {
+
+		when(requestCreator.search()).thenReturn(searchRequestConverter);
+		SearchRequest searchRequest = new SearchRequest("someindex") //
+				.source(new SearchSourceBuilder().query(QueryBuilders.matchAllQuery()));
+
+		ReactiveElasticsearchClient client = new DefaultReactiveElasticsearchClient(hostProvider, requestCreator) {
 			@Override
 			public Mono<ResponseSpec> execute(ReactiveElasticsearchClientCallback callback) {
 				return Mono.empty();
 			}
 		};
-	}
-
-	@Test
-	void shouldSetAppropriateRequestParametersOnCount() {
-
-		SearchRequest searchRequest = new SearchRequest("someindex") //
-				.source(new SearchSourceBuilder().query(QueryBuilders.matchAllQuery()));
 
 		client.count(searchRequest).as(StepVerifier::create).verifyComplete();
 
@@ -78,5 +81,34 @@ class DefaultReactiveElasticsearchClientTest {
 		assertThat(source.size()).isEqualTo(0);
 		assertThat(source.trackTotalHitsUpTo()).isEqualTo(TRACK_TOTAL_HITS_ACCURATE);
 		assertThat(source.fetchSource()).isEqualTo(FetchSourceContext.DO_NOT_FETCH_SOURCE);
+	}
+
+	@Test // #1712
+	@DisplayName("should throw ElasticsearchStatusException on server 5xx with empty body")
+	void shouldThrowElasticsearchStatusExceptionOnServer5xxWithEmptyBody() {
+
+		when(hostProvider.getActive(any())).thenReturn(Mono.just(webClient));
+		WebClient.RequestBodyUriSpec requestBodyUriSpec = mock(WebClient.RequestBodyUriSpec.class);
+		when(requestBodyUriSpec.uri((Function<UriBuilder, URI>) any())).thenReturn(requestBodyUriSpec);
+		when(requestBodyUriSpec.attribute(any(), any())).thenReturn(requestBodyUriSpec);
+		when(requestBodyUriSpec.headers(any())).thenReturn(requestBodyUriSpec);
+		when(webClient.method(any())).thenReturn(requestBodyUriSpec);
+		when(requestBodyUriSpec.exchangeToMono(any())).thenAnswer(invocationOnMock -> {
+			Function<ClientResponse, ? extends Mono<?>> responseHandler = invocationOnMock.getArgument(0);
+			ClientResponse clientResponse = mock(ClientResponse.class);
+			when(clientResponse.statusCode()).thenReturn(HttpStatus.SERVICE_UNAVAILABLE);
+			ClientResponse.Headers headers = mock(ClientResponse.Headers.class);
+			when(headers.contentType()).thenReturn(Optional.empty());
+			when(clientResponse.headers()).thenReturn(headers);
+			when(clientResponse.body(any())).thenReturn(Mono.empty());
+			return responseHandler.apply(clientResponse);
+		});
+
+		ReactiveElasticsearchClient client = new DefaultReactiveElasticsearchClient(hostProvider, requestCreator);
+
+		client.get(new GetRequest("42")) //
+				.as(StepVerifier::create) //
+				.expectError(ElasticsearchStatusException.class) //
+				.verify(); //
 	}
 }

